runtime/lua/vim/lsp/_tagfunc.lua

```diff
@@ -35,23 +37,33 @@ local function query_definition(pattern)
     table.insert(results, mk_tag_item(pattern, range, uri, offset_encoding))
   end
 
-  for client_id, lsp_results in pairs(assert(results_by_client)) do
-    local client = lsp.get_client_by_id(client_id)
-    local offset_encoding = client and client.offset_encoding or 'utf-16'
-    local result = lsp_results.result or {}
-    if result.range then -- Location
-      add(result.range, result.uri, offset_encoding)
-    else
-      result = result --[[@as (lsp.Location[]|lsp.LocationLink[])]]
-      for _, item in pairs(result) do
-        if item.range then -- Location
-          add(item.range, item.uri, offset_encoding)
-        else -- LocationLink
-          add(item.targetSelectionRange, item.targetUri, offset_encoding)
+  local remaining = #clients
+  for _, client in ipairs(clients) do
+    ---@param result nil|lsp.Location|lsp.Location[]|lsp.LocationLink[]
+    local function on_response(_, result)
+      if result then
+        local encoding = client.offset_encoding
+        -- single Location
+        if result.range then
+          add(result.range, result.uri, encoding)
+        else
+          for _, location in ipairs(result) do
+            if location.range then -- Location
+              add(location.range, location.uri, encoding)
+            else -- LocationLink
+              add(location.targetSelectionRange, location.targetUri, encoding)
+            end
+          end
         end
       end
+      remaining = remaining - 1
     end
+    local params = util.make_position_params(win, client.offset_encoding)
+    client.request(ms.textDocument_definition, params, on_response, bufnr)
   end
+  vim.wait(1000, function()
```

Comment:
A comment why this `wait` is needed would be helpful.

Reply:
It is because this function no longer uses `lsp.buf_request_sync` (which calls `vim.wait` internally), and this function requires the results to be returned synchronously.

Every other LSP feature has to be aware of the async nature (specifically `client.request`), so I'd be wary about commenting too much on this. 

Since "the new way" of doing things has similar patterns, I'm hoping we can generalize this once we've eliminated "the old way".